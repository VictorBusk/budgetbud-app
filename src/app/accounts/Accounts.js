import React, {PureComponent} from 'react';
import {FlatList} from 'react-native';
import {connect} from 'react-redux';
import _ from 'lodash';
import {
    AppHeader,
    Separator
} from "../../components/index";
import {
    Container,
    ListItem,
    Body,
    Right,
    Text,
    Spinner,
    CheckBox,
    Button
} from 'native-base';
import I18n from "../../strings/i18n";
import {
    button,
    text,
    list,
    container, color
} from "../../style/index";
import {showWarningToast} from "../../helpers/index";
import {
    accountsSelected,
    getAccounts,
    linkAccounts,
    resetAccountsError
} from "../../redux/actions";

class Accounts extends PureComponent {
    componentWillMount() {
        if (!this.props.accountsInitialized)
            this.props.getAccounts();
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.accountsError) {
            showWarningToast(nextProps.accountsError);
            this.props.resetAccountsError();
        }
    }

    onSavePress = () => {
        this.props.linkAccounts(this.props.linkedAccounts, () => {
            this.props.navigation.pop();
        });
    };

    onCheckBoxPress = ({accountID}) => {
        let tmp = this.props.linkedAccounts;

        if (tmp.includes(accountID)) {
            tmp.splice(tmp.indexOf(accountID), 1)
        } else {
            tmp.push(accountID);
        }

        this.props.accountsSelected(tmp);
    };

    render() {
        return (
            <Container style={container.signedInContainer}>
                <AppHeader headerText={I18n.t('accountsHeader')}
                           showBackButton={true}
                           onLeftButtonPress={() => {
                               if (!this.props.linkLoading)
                                   this.props.navigation.pop()
                           }}
                />

                <Container style={{justifyContent: 'center'}}>
                    {this.props.accountsLoading ? (
                        <Spinner style={{
                            alignItems: 'center',
                            justifyContent: 'center'
                        }} color='#1c313a'/>) : (
                        <FlatList
                            data={this.props.accountItems}
                            renderItem={this.renderItem}
                        />
                    )}
                </Container>

                <Separator/>

                <Button rounded
                        onPress={() => {
                            if (!this.props.linkLoading || !this.props.accountsLoading) {
                                this.onSavePress()
                            }
                        }}
                        style={[button.defaultButton, color.button]}
                >
                    {this.props.linkLoading ? (
                        <Spinner color='#D0D0D0'/>) : (
                        <Text style={text.submitButtonText}>{I18n.t('accountsSaveButton')}</Text>
                    )}
                </Button>
            </Container>
        );
    }

    renderItem = ({item}) => {
        const checked = this.props.linkedAccounts.includes(item.accountID);

        return (
            <ListItem>
                <Body>
                <Text style={color.text}>{item.name}</Text>
                </Body>
                <Right style={list.listItemCheckBoxPadding}>
                    <CheckBox
                        style={checked ? color.checkboxChecked : color.checkboxUnchecked}
                        checked={checked}
                        onPress={() => this.onCheckBoxPress(item)}
                    />
                </Right>
            </ListItem>
        );
    }
}

const mapStateToProps = ({account}) => {
    const {
        accounts,
        linkedAccounts,
        accountsLoading,
        linkLoading,
        accountsError,
        accountsInitialized
    } = account;

    const accountItems = _.map(accounts, (item, key) => {
        return {name: item.name, accountID: item.id, key: key};
    });

    return {
        accountItems,
        linkedAccounts,
        accountsLoading,
        linkLoading,
        accountsError,
        accountsInitialized
    };
};

const mapDispatchToProps = {
    getAccounts,
    linkAccounts,
    accountsSelected,
    resetAccountsError
};

export default connect(mapStateToProps, mapDispatchToProps)(Accounts);
