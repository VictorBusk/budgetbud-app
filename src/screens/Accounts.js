import React, {PureComponent} from 'react';
import {FlatList} from 'react-native';
import {connect} from 'react-redux';
import {
    AppHeader,
    Separator
} from "../components/";
import {
    Container,
    ListItem,
    Body,
    Right,
    Text,
    Spinner,
    CheckBox,
    Button
} from 'native-base';
import I18n from "../strings/i18n";
import {
    getAccounts,
    linkAccounts,
    accountsSelected
} from "../actions";
import {button, text, list} from "../style";

class Accounts extends PureComponent {
    componentWillMount() {
        this.props.getAccounts();
    }

    onSavePress = () => {
        this.props.linkAccounts(this.props.selectedAccounts);
    };

    onCheckBoxPress = ({accountID}) => {
        let tmp = this.props.selectedAccounts;

        if (tmp.includes(accountID)) {
            tmp.splice(tmp.indexOf(accountID), 1)
        } else {
            tmp.push(accountID);
        }

        this.props.accountsSelected(tmp);
    };

    render() {
        return (
            <Container>
                <Container>
                    <AppHeader headerText={I18n.t('accountsHeader')}
                               showBackButton={true}
                               onLeftButtonPress={() => this.props.navigation.pop()}/>

                    <Container style={{flex: 4, justifyContent: 'center'}}>
                        {this.props.accountsLoading ? (
                            <Spinner style={{
                                alignItems: 'center',
                                justifyContent: 'center'
                            }} color='#1c313a'/>) : (
                            <FlatList
                                data={this.props.accountItems}
                                renderItem={this.renderItem}
                            />
                        )}
                    </Container>

                    <Separator/>

                    <Button rounded
                            onPress={() => {if (!this.props.linkLoading) {this.onSavePress()}}}
                            style={button.defaultButton}
                    >
                        {this.props.linkLoading ? (
                            <Spinner color='#D0D0D0'/>) : (
                            <Text style={text.submitButtonText}>{I18n.t('accountsSaveButton')}</Text>
                        )}
                    </Button>
                </Container>
            </Container>
        );
    }

    renderItem = ({item}) => {
        return (
            <ListItem>
                <Body>
                <Text>{item.name}</Text>
                </Body>
                <Right style={list.listItemCheckBoxPadding}>
                    <CheckBox
                        style={{borderColor: '#777777'}}
                        checked={this.props.selectedAccounts.includes(item.accountID)}
                        onPress={() => this.onCheckBoxPress(item)}
                    />
                </Right>
            </ListItem>
        );
    }
}

const mapStateToProps = ({account}) => {
    const {
        accounts,
        selectedAccounts,
        accountsLoading,
        linkLoading
    } = account;

    const accountItems = _.map(accounts, (item, key) => {
        return {name: item.name, accountID: item.id, key: key};
    });

    return {
        accountItems,
        selectedAccounts,
        accountsLoading,
        linkLoading
    };
};

const mapDispatchToProps = {
    getAccounts,
    linkAccounts,
    accountsSelected,
};

export default connect(mapStateToProps, mapDispatchToProps)(Accounts);
