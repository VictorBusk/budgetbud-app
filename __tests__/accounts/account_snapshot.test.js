import React from 'react';
import configureStore from 'redux-mock-store';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Accounts from "../../src/app/accounts/Accounts";

Enzyme.configure({adapter: new Adapter()});
import thunk from 'redux-thunk';
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Testing Accounts', () => {
    it('renders initially as expected', () => {
        const state = {
            accounts: [],
            linkedAccounts: [],
            accountsLoading: false,
            linkLoading: false,
            accountsError: '',
            accountsInitialized: false,
        };

        const wrapper = shallow(
            <Accounts/>,
            {context: {store: mockStore({account: state})}},
        );
        expect(wrapper.dive()).toMatchSnapshot();
    });
});
