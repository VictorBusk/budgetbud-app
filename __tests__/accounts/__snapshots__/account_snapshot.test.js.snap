// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Accounts renders initially as expected 1`] = `
<Styled(Container)
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
>
  <AppHeader
    headerText="Accounts"
    infoText="Here you can choose which accounts you want expenses from. Press [Save] to save the changes"
    onLeftButtonPress={[Function]}
    showBackButton={true}
  />
  <Styled(Container)
    style={
      Object {
        "justifyContent": "center",
      }
    }
  >
    <FlatList
      data={Array []}
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </Styled(Container)>
  <Separator />
  <Styled(Button)
    onPress={[Function]}
    rounded={true}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "height": 40,
          "justifyContent": "center",
          "marginBottom": 20,
          "marginTop": 20,
          "width": "90%",
        },
        Object {
          "backgroundColor": "#00263A",
        },
      ]
    }
  >
    <Styled(Text)
      style={
        Object {
          "alignSelf": "center",
          "color": "white",
          "fontWeight": "600",
        }
      }
    >
      Save
    </Styled(Text)>
  </Styled(Button)>
</Styled(Container)>
`;

exports[`Testing SignIn renders as expected 1`] = `
<Styled(Container)
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
>
  <AppHeader
    headerText="Accounts"
    infoText="Here you can choose which accounts you want expenses from. Press [Save] to save the changes"
    onLeftButtonPress={[Function]}
    showBackButton={true}
  />
  <Styled(Container)
    style={
      Object {
        "justifyContent": "center",
      }
    }
  >
    <FlatList
      data={Array []}
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </Styled(Container)>
  <Separator />
  <Styled(Button)
    onPress={[Function]}
    rounded={true}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "height": 40,
          "justifyContent": "center",
          "marginBottom": 20,
          "marginTop": 20,
          "width": "90%",
        },
        Object {
          "backgroundColor": "#00263A",
        },
      ]
    }
  >
    <Styled(Text)
      style={
        Object {
          "alignSelf": "center",
          "color": "white",
          "fontWeight": "600",
        }
      }
    >
      Save
    </Styled(Text)>
  </Styled(Button)>
</Styled(Container)>
`;
